### PR TITLE
add EBS-CSI warnings

### PR DESCRIPTION
**Pull request by AjayTripathy**
## What does this PR change?
Add warnings to notes on 

## Does this PR rely on any other PRs?
No

## How does this PR impact users? (This is the kind of thing that goes in release notes!)
Helps users debug EKS PV issues

## How was this PR tested?
Produces the message on an EKS cluster. 
```
atripathy@Ajays-MBP cost-analyzer-helm-chart % helm install kubecost ./cost-analyzer -n ajaytest
NAME: kubecost
LAST DEPLOYED: Sun Jan 22 16:13:03 2023
NAMESPACE: ajaytest
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
--------------------------------------------------Kubecost has been successfully installed.

WARNING: ON EKS v1.23+ INSTALLATION OF EBS-CSI DRIVER IS REQUIRED TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html

We detected a EBS-CSI Driver, but you should check that Kubecost's PersistentVolume has been installed correctly

Please allow 5-10 minutes for Kubecost to gather metrics.

If you have configured cloud-integrations, it can take up to 48 hours for cost reconciliation to occur.

When using Durable storage (Enterprise Edition), please allow up to 4 hours for data to be collected and the UI to be healthy.

When pods are Ready, you can enable port-forwarding with the following command:

    kubectl port-forward --namespace ajaytest deployment/kubecost-cost-analyzer 9090

Next, navigate to http://localhost:9090 in a web browser.

Having installation issues? View our Troubleshooting Guide at http://docs.kubecost.com/troubleshoot-install
```

## Have you made an update to documentation?
This is documentation, more or less.

### CODE REVIEW

File: cost-analyzer/templates/NOTES.txt [hunk 1]
```diff
@@ -1,9 +1,35 @@
...
+
+WARNING: ON EKS v1.23+ INSTALLATION OF EBS-CSI DRIVER IS REQUIRED TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
+
+{{ if (and $EBSCSINotExists $PVNotExists) -}}
+
+ERROR: MISSING EBS-CSI DRIVER WHICH IS REQUIRED ON EKS v1.23+ TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
+
+{{ else if (and $EBSCSINotExists (not $PVNotExists)) -}}
+
+ERROR: MISSING EBS-CSI DRIVER WHICH IS REQUIRED ON EKS v1.23+ TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
```
Comment (dwbrown2) [inline cost-analyzer/templates/NOTES.txt]: Should we stop install process on such a clear error? @linhlam-kc @kwombach12

File: cost-analyzer/templates/NOTES.txt [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
 {{- $servicePort := .Values.service.port | default 9090 -}}
 Kubecost has been successfully installed.
 
+{{ if (and $isEKS $isGT22) -}}
+
+WARNING: ON EKS v1.23+ INSTALLATION OF EBS-CSI DRIVER IS REQUIRED TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
+
+{{ if (and $EBSCSINotExists $PVNotExists) -}}
+
+ERROR: MISSING EBS-CSI DRIVER WHICH IS REQUIRED ON EKS v1.23+ TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
```
Comment (dwbrown2) [inline cost-analyzer/templates/NOTES.txt]: Is a message needed in this scenario?

File: cost-analyzer/templates/NOTES.txt [hunk 3]
```diff
@@ -1,9 +1,35 @@
...
+
+{{ else if (and $EBSCSINotExists (not $PVNotExists)) -}}
+
+ERROR: MISSING EBS-CSI DRIVER WHICH IS REQUIRED ON EKS v1.23+ TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html
+
+{{ else if (not $EBSCSINotExists) -}}
+
+We detected a EBS-CSI Driver, but you should check that Kubecost's PersistentVolume has been installed correctly
```
Comment (dwbrown2) [inline cost-analyzer/templates/NOTES.txt]: What's the value in this message? Is it needed?

File: cost-analyzer/templates/NOTES.txt (hunk 1 above)
Comment (AjayTripathy) [reply cost-analyzer/templates/NOTES.txt]: We can do that with a "fail" function, just leaving as a note once we decide:

https://stackoverflow.com/questions/61496709/helm-template-how-raise-exception-in-in-helm-function

File: cost-analyzer/templates/NOTES.txt (hunk 2 above)
Comment (AjayTripathy) [reply cost-analyzer/templates/NOTES.txt]: You flip the conditional here? This is the case where you don't have any PVs, which is more expected when there's not EBS CSI driver. Above is the less expected one imo?

File: cost-analyzer/templates/NOTES.txt (hunk 3 above)
Comment (AjayTripathy) [reply cost-analyzer/templates/NOTES.txt]: I don't know. I would want an opinion on whether the ebs CSI driver 's name is often changed or if it is installed in other namespaces sometimes.

File: cost-analyzer/templates/NOTES.txt [hunk 4]
```diff
@@ -1,9 +1,35 @@
 
 
 --------------------------------------------------
+{{- $node := (lookup "v1" "Node" "" "") }}
+{{- $isEKS := (regexMatch ".*eks.*" (.Capabilities.KubeVersion | quote) )}}
+{{- $isGT22 := (semverCompare ">=1.23-0" .Capabilities.KubeVersion.GitVersion) }}
+{{- $PVNotExists := (empty (lookup "v1" "PersistentVolume" "" "")) }}
+{{- $EBSCSINotExists := (empty (lookup "apps/v1" "Deployment" "kube-system" "ebs-csi-controller")) }}
```
Comment (dwbrown2) [inline cost-analyzer/templates/NOTES.txt:8]: would you ever not have namespace access here? if you don't have access, what happens?

File: cost-analyzer/templates/NOTES.txt (hunk 1 above)
Comment (linhlam-kc) [reply cost-analyzer/templates/NOTES.txt]: IMO, we should not put a hard stop here but add more details for the recommended action. E.g:
`TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html OR TRY OUR FOLLOWING DOCUMENTATION HERE: https://docs.kubecost.com/install-and-configure/install/provider-installations/aws-eks-cost-monitoring#prerequisites` . If there is feedback from users that they absolutely wants to have a hard stop here, we can implement it later.

Comment (dwbrown2) [reply cost-analyzer/templates/NOTES.txt]: My only concern is that users miss this Error message and are still confused... maybe we make `ERROR` bold & red?

**Comment by linhlam-kc** [comment]
Adding some deployment tests from my existing cluster:

# TEST CASES
## Install on EKS 1.23, no EBS CSI driver:

```
helm install kubecost cost-analyzer-helm-chart/cost-analyzer/ -n kubecost --create-namespace
W0123 11:44:35.350602   20709 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0123 11:44:35.508768   20709 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0123 11:44:45.859583   20709 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0123 11:44:45.865055   20709 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
NAME: kubecost
LAST DEPLOYED: Mon Jan 23 11:44:33 2023
NAMESPACE: kubecost
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
--------------------------------------------------Kubecost has been successfully installed.

WARNING: ON EKS v1.23+ INSTALLATION OF EBS-CSI DRIVER IS REQUIRED TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html

ERROR: MISSING EBS-CSI DRIVER WHICH IS REQUIRED ON EKS v1.23+ TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html

Please allow 5-10 minutes for Kubecost to gather metrics.

If you have configured cloud-integrations, it can take up to 48 hours for cost reconciliation to occur.

When using Durable storage (Enterprise Edition), please allow up to 4 hours for data to be collected and the UI to be healthy.

When pods are Ready, you can enable port-forwarding with the following command:

    kubectl port-forward --namespace kubecost deployment/kubecost-cost-analyzer 9090

Next, navigate to http://localhost:9090 in a web browser.

Having installation issues? View our Troubleshooting Guide at http://docs.kubecost.com/troubleshoot-install
```
## Install on EKS 1.23, EBS CSI driver is installed:
```
helm install kubecost cost-analyzer-helm-chart/cost-analyzer/ -n kubecost --create-namespace
W0123 12:16:53.144948   21636 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0123 12:16:53.301684   21636 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0123 12:17:03.779960   21636 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0123 12:17:03.784079   21636 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
NAME: kubecost
LAST DEPLOYED: Mon Jan 23 12:16:51 2023
NAMESPACE: kubecost
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
--------------------------------------------------Kubecost has been successfully installed.

WARNING: ON EKS v1.23+ INSTALLATION OF EBS-CSI DRIVER IS REQUIRED TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html

We detected a EBS-CSI Driver, but you should check that Kubecost's PersistentVolume has been installed correctly

Please allow 5-10 minutes for Kubecost to gather metrics.

If you have configured cloud-integrations, it can take up to 48 hours for cost reconciliation to occur.

When using Durable storage (Enterprise Edition), please allow up to 4 hours for data to be collected and the UI to be healthy.

When pods are Ready, you can enable port-forwarding with the following command:

    kubectl port-forward --namespace kubecost deployment/kubecost-cost-analyzer 9090

Next, navigate to http://localhost:9090 in a web browser.

Having installation issues? View our Troubleshooting Guide at http://docs.kubecost.com/troubleshoot-install
```

File: cost-analyzer/templates/NOTES.txt (hunk 3 above)
Comment (linhlam-kc) [reply cost-analyzer/templates/NOTES.txt]: In my test on EKS 1.23 and EBS CSI is installed. This message still appears. I think it is a little bit confusing to users/customers. In most cases the EBS CSI driver's name should not be changed or it is deployed outside of kube-system as it is not AWS recommendation.

File: cost-analyzer/templates/NOTES.txt (hunk 1 above)
Comment (linhlam-kc) [reply cost-analyzer/templates/NOTES.txt]: Afaik, the NOTES is plain text, so I'm not sure if we can format the texts. 

On the note of whether we should put a hard stop. My concern is that putting a hard stop may block users from using Kubecost, if they have EBS driver installed in non-standard way or some use cases that are not covered by our logic. It looks like an one-way-door to me that may require us to release a hot patch or users have to wait until the next release of our chart in that case.

In addition, I see this as an Amazon EKS limitation but not our product limitation.

**Review by linhlam-kc — COMMENTED** [review]
It works as expected in the test cases I shared in my previous comment. Just 1 suggestion that we may consider is to add instructions to enable EBS CSI driver from our documentation at [LINK](https://docs.kubecost.com/install-and-configure/install/provider-installations/aws-eks-cost-monitoring#prerequisites) . Amazon documentation could be hard to follow for some users just looking for quick instructions to install EBS CSI driver.

File: cost-analyzer/templates/NOTES.txt (hunk 3 above)
Comment (AjayTripathy) [reply cost-analyzer/templates/NOTES.txt]: OK, I can remove! Thanks for the insight.

File: cost-analyzer/templates/NOTES.txt (hunk 4 above)
Comment (AjayTripathy) [reply cost-analyzer/templates/NOTES.txt:8]: IMO you should get the warning.

But I think after a quick test this actually will block install. Let me see if it's quick to add a try-catch of some sort here.

Comment (AjayTripathy) [reply cost-analyzer/templates/NOTES.txt:8]: Yep, it will block install:
```
The lookup function uses Helm's existing Kubernetes connection configuration to query Kubernetes. If any error is returned when interacting with calling the API server (for example due to lack of permission to access a resource), helm's template processing will fail.
```

**Comment by linhlam-kc** [comment]
# TEST CASES
## Install on EKS 1.23, no EBS CSI driver:

```
helm install kubecost1 cost-analyzer/ -n kubecost1
W0130 08:28:15.738851   40980 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0130 08:28:15.876241   40980 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0130 08:28:23.140901   40980 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0130 08:28:23.145740   40980 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
NAME: kubecost1
LAST DEPLOYED: Mon Jan 30 08:28:14 2023
NAMESPACE: kubecost1
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
--------------------------------------------------Kubecost has been successfully installed.

WARNING: ON EKS v1.23+ INSTALLATION OF EBS-CSI DRIVER IS REQUIRED TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.kubecost.com/install-and-configure/install/provider-installations/aws-eks-cost-monitoring#prerequisites

ERROR: MISSING EBS-CSI DRIVER WHICH IS REQUIRED ON EKS v1.23+ TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.kubecost.com/install-and-configure/install/provider-installations/aws-eks-cost-monitoring#prerequisites

Please allow 5-10 minutes for Kubecost to gather metrics.

If you have configured cloud-integrations, it can take up to 48 hours for cost reconciliation to occur.

When using Durable storage (Enterprise Edition), please allow up to 4 hours for data to be collected and the UI to be healthy.

When pods are Ready, you can enable port-forwarding with the following command:

    kubectl port-forward --namespace kubecost1 deployment/kubecost1-cost-analyzer 9090

Next, navigate to http://localhost:9090 in a web browser.

Having installation issues? View our Troubleshooting Guide at http://docs.kubecost.com/troubleshoot-install
```
## Install on EKS 1.23, EBS CSI driver is installed:
```
helm install kubecost1 cost-analyzer/ -n kubecost1
W0130 08:34:24.254204   41172 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0130 08:34:24.392956   41172 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0130 08:34:31.432206   41172 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
W0130 08:34:31.435562   41172 warnings.go:70] policy/v1beta1 PodSecurityPolicy is deprecated in v1.21+, unavailable in v1.25+
NAME: kubecost1
LAST DEPLOYED: Mon Jan 30 08:34:22 2023
NAMESPACE: kubecost1
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
--------------------------------------------------Kubecost has been successfully installed.

WARNING: ON EKS v1.23+ INSTALLATION OF EBS-CSI DRIVER IS REQUIRED TO MANAGE PERSISTENT VOLUMES. LEARN MORE HERE: https://docs.kubecost.com/install-and-configure/install/provider-installations/aws-eks-cost-monitoring#prerequisites

Please allow 5-10 minutes for Kubecost to gather metrics.

If you have configured cloud-integrations, it can take up to 48 hours for cost reconciliation to occur.

When using Durable storage (Enterprise Edition), please allow up to 4 hours for data to be collected and the UI to be healthy.

When pods are Ready, you can enable port-forwarding with the following command:

    kubectl port-forward --namespace kubecost1 deployment/kubecost1-cost-analyzer 9090

Next, navigate to http://localhost:9090 in a web browser.

Having installation issues? View our Troubleshooting Guide at http://docs.kubecost.com/troubleshoot-install
```

**Review by linhlam-kc — APPROVED** [review]
LGTM !

**Comment by dwbrown2** [comment]
Thanks for testing, @linhlam-kc!

Do we need to show both error + warning on the first case? Also, should we link to our docs over the AWS docs?

@AjayTripathy can we change to red color when Error is present?

**Comment by linhlam-kc** [comment]
> Thanks for testing, @linhlam-kc!
> 
> Do we need to show both error + warning on the first case? Also, should we link to our docs over the AWS docs?
> 
> @AjayTripathy can we change to red color when Error is present?

Hi @dwbrown2 , I think it is fine to link to our doc where we also have the link to AWS doc that explained the requirement. I think it can help to reduce to customer confusion as they need to know how to fix the issue first before getting to know more about the details. On the first case, I think the WARN can be removed with the information from the ERROR

**Comment by AjayTripathy** [comment]
> Thanks for testing, @linhlam-kc!

> Do we need to show both error + warning on the first case? 

Yes, IMO. I'd rather have the warning across all EKS 1.23 clusters for a few releases so people can understand and install the driver.

>Also, should we link to our docs over the AWS docs?

Per Linh's feedback I'm linking to our docs.

> @AjayTripathy can we change to red color when Error is present?

I don't see a good way to do this.
